connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

```diff
@@ -666,9 +667,10 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
         testUtils.sendMessages(topic2, Array("6"))
       },
       StartStream(),
-      ExpectFailure[IllegalStateException](e => {
+      ExpectFailure[SparkException](e => {
+        assert(e.asInstanceOf[SparkThrowable].getErrorClass === "INTERNAL_ERROR")
         // The offset of `topic2` should be changed from 2 to 1
-        assert(e.getMessage.contains("was changed from 2 to 1"))
+        assert(e.getCause.getMessage.contains("was changed from 2 to 1"))
```

Comment:
I'm looking into this test and feels like this should be a user-facing error as users can trigger it with valid queries. @HeartSaVioR what do you think?

Reply:
It's a convention in Spark (and many other Java programs) that `IllegalStateException` means a bug. If this is not a bug, I think streaming code should not throw `IllegalStateException` in this case.

Reply:
https://docs.oracle.com/javase/8/docs/api/java/lang/IllegalStateException.html

> Signals that a method has been invoked at an illegal or inappropriate time. In other words, the Java environment or Java application is not in an appropriate state for the requested operation.

It's arguable what is the "state" of the application. If the state of the application is dependent on the external system (like this), this statement sounds to me to be valid.

If Spark project wants to restrict the usage of IllegalStateException to only denote the possible internal error then that is fair, but would be great if we do not rely on "convention" (that's a magic word everyone can claim the different things with the same word) and explicitly mention it.

Reply:
> Signals that a method has been invoked at an illegal or inappropriate time.

This means a bug in any product, right? We should not expose these low-level errors to end-users, same to NPE, index out of bounds error, etc.

Reply:
If Spark is a simple Java library, then this is fine. The caller should handle these errors property when building a product.

Reply:
What makes the time as "illegal" or "inappropriate" is the matter. It does not only bind to the bug of the application. The definition is not strict enough - if we call readTable whereas concurrent operation on non-atomic drop-and-recreate table is happening, the time is "conditionally" "inappropriate".

For sure, we can be strict on the project's policy to follow the convention you mentioned (probably define new more-specific exception(s) if needed). For Kafka data source, we'll need some time to sort out on this.

Reply:
I just realized this is included in Spark 3.3.0 instead of 3.4.0... I thought we have sufficient time to look into, but it doesn't seem so unfortunately...

Can we check how the error message has changed for the test in user facing, and decide whether it is a breaking change or not? If the error class framework hides the details since it's considered as an internal error, this is a huge breaking change since we give a guidance in the error message.

Reply:
And we may need to make this clear to the interface contract on data source - now we are applying the error class framework even the exception comes from the data source. The convention, only applies in Spark project itself.

Reply:
The error is simply wrapped and the original error is kept in causedBy, so this is not a big problem. The bigger problem is how to define internal errors. We picked some common java exceptions that can be treated as internal errors: when end users see these errors from Spark, it indicates a bug.

For third-party Spark plugins, I think all errors from them are internal errors? end-users can do nothing to work around these errors. cc @srielau

Reply:
Just to close the loop; we decided to deal with this issue via https://issues.apache.org/jira/browse/SPARK-39412.